File: Tests/SentryTests/Networking/TestNSURLRequestBuilder.m
#import "TestNSURLRequestBuilder.h"

NS_ASSUME_NONNULL_BEGIN

@interface
TestNSURLRequestBuilder ()

@property (nonatomic, strong) SentryNSURLRequestBuilder *builder;
@property (nonatomic, strong) NSError *error;

@end

@implementation TestNSURLRequestBuilder

- (instancetype)init
{
    if (self = [super init]) {
        self.builder = [[SentryNSURLRequestBuilder alloc] init];
    }
    return self;
}

- (NSURLRequest *)createEnvelopeRequest:(SentryEnvelope *)envelope
                                    dsn:(SentryDsn *)dsn
                       didFailWithError:(NSError *_Nullable *_Nullable)error
{
    NSURLRequest *request = [self.builder createEnvelopeRequest:envelope
                                                            dsn:dsn
                                               didFailWithError:error];
    if (self.shouldFailWithError) {
        self.error = [[NSError alloc] initWithDomain:@"TestErrorDomain" code:12 userInfo:nil];
        *error = self.error;
    }
    return request;
}

- (NSURLRequest *)createEnvelopeRequest:(SentryEnvelope *)envelope
                                    url:(NSURL *)url
                       didFailWithError:(NSError *_Nullable *_Nullable)error
{
    NSURLRequest *request = [self.builder createEnvelopeRequest:envelope
                                                            url:url
                                               didFailWithError:error];
    if (self.shouldFailWithError) {
        self.error = [[NSError alloc] initWithDomain:@"TestErrorDomain" code:12 userInfo:nil];
        *error = self.error;
    }
    return request;
}

@end

NS_ASSUME_NONNULL_END

File: Tests/SentryTests/Networking/TestNSURLRequestBuilder.swift
#if SWIFT_PACKAGE
@_spi(Private) @testable import SentrySwift
import _SentryPrivate
import SentryObjCInternal
import SentryTestsObjCHelpers
#else
@testable @_spi(Private) import Sentry
#endif

final class TestNSURLRequestBuilder: SentryNSURLRequestBuilder {
    var shouldFailWithError: Bool = false
    var error: NSError?
    let builder = SentryNSURLRequestBuilder()
    
    public override func createEnvelopeRequest(_ envelope: SentryEnvelope, dsn: SentryDsn) throws -> URLRequest {
        let request = try builder.createEnvelopeRequest(envelope, dsn: dsn)
        if shouldFailWithError {
            let error = NSError(domain: "TestErrorDomain", code: 12)
            self.error = error
            throw error
        }
        return request
    }
    
    public override func createEnvelopeRequest(_ envelope: SentryEnvelope, url: URL) throws -> URLRequest {
        let request = try builder.createEnvelopeRequest(envelope, url: url)
        if shouldFailWithError {
            let error = NSError(domain: "TestErrorDomain", code: 12)
            self.error = error
            throw error
        }
        return request
    }
        
}
